torch_xla/csrc/helpers.cpp

```diff
@@ -92,12 +94,22 @@ std::pair<xla::XlaOp, xla::XlaOp> XlaHelpers::PromoteValues(
   if (type1 == type2) {
     return std::pair<xla::XlaOp, xla::XlaOp>(op1, op2);
   }
-  if (type1 == xla::PrimitiveType::F32) {
+  xla::int64 size1 = xla::ShapeUtil::ByteSizeOfPrimitiveType(type1);
```

Comment:
What happens if promotion happens as part of the gradient computation and the weights are BF16? For experimentation, this is fine, but to make it robust looks like we'd need a pass to insert additional conversions back to BF16.

Reply:
We call this API explicitly to make it possible to perform an operation between two operands with possible different type.
In theory every at::Float tensor we create with that flag on, will result in BF16.
XLA does the same (promote to higher FP precision) if we do not do this, anyway.

tensorflow/compiler/xla/service/shape_inference.cc:865

But sure, we can do more investigation and see the outputs.
So far I have verified that MNIST and Resnt18/CIFAR10 train correctly with that options enabled.